parser/sqlm/parse.go

```diff
@@ -0,0 +1,407 @@
+package sqlm
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+)
+
+type clause struct {
+	isExec bool
+
+	Select string
+	From   string
+	Joins  []string
+
+	Full string
+}
+
+type keyword string
+
+func (k keyword) eq(s string) bool {
+	return strings.ToUpper(s) == string(k)
+}
+
+var (
+	_select = keyword("SELECT")
+	_from   = keyword("FROM")
+	_join   = keyword("JOIN")
+	_on     = keyword("ON")
+	_count  = keyword("COUNT")
+
+	// Keywords after 'FROM' clause.
+	afterFroms = map[string]struct{}{
+		"LEFT": {}, "RIGHT": {}, "INNER": {}, "JOIN": {},
+		"WHERE": {}, "ORDER": {}, "LIMIT": {}, "GROUP": {},
+	}
+)
+
+// split extracts different components in the sql, that is,
+// different clauses. We mainly focus on the SELECT, FROM,
+// and JOIN clauses of the query sql for subsequent analysis
+// and template data generation.
+func split(sql, raw string) (*clause, error) {
+	tokens := strings.Fields(sql)
+	if len(tokens) == 0 {
+		return nil, errors.New("sql is empty")
+	}
+	// First part: SELECT clause.
+	master := tokens[0]
+	if !_select.eq(master) {
+		// No query sql, no need to parse.
+		return &clause{isExec: true}, nil
+	}
+	tokens = tokens[1:]
+	var selectTokens []string
+	var idx int
+	for {
+		if idx == len(tokens) {
+			var last string
+			if idx-1 >= 0 {
+				last = tokens[idx-1]
+			}
+			// FROM not found
+			return nil, &Error{
+				full:  raw,
+				wrong: last,
+				desc:  "FROM not found in sql",
+			}
+		}
+
+		t := tokens[idx]
+		if _from.eq(t) {
+			break
+		}
+
+		selectTokens = append(selectTokens, t)
+		idx++
+	}
+	if len(selectTokens) == 0 {
+		return nil, &Error{
+			full:  raw,
+			wrong: master,
+			desc:  "SELECT is empty",
+		}
+	}
+	cs := new(clause)
+	cs.Full = raw
+	cs.Select = strings.Join(selectTokens, " ")
+	fromToken := tokens[idx]
+	tokens = tokens[idx+1:]
+
+	// Second part: FROM clause.
+	var fromTokens []string
+	idx = 0
+	for {
+		if idx == len(tokens) {
+			// sql ends.
+			break
+		}
+		t := tokens[idx]
+		if _, ok := afterFroms[t]; ok {
+			// meet end keyword.
+			break
+		}
+		fromTokens = append(fromTokens, t)
+		idx++
+	}
+	if len(fromTokens) == 0 {
+		return nil, &Error{
+			full:  sql,
+			wrong: fromToken,
+			desc:  "FROM clause is empty",
+		}
+	}
+	cs.From = strings.Join(fromTokens, " ")
+	tokens = tokens[idx:]
+
+	// Third part: JOIN clause.
+	idx = 0
+	for {
+		if idx == len(tokens) {
+			break
+		}
+		t := tokens[idx]
+		if !_join.eq(t) {
+			idx++
+			continue
+		}
+		var joinTokens []string
+		for {
+			idx++
+			if idx == len(tokens) {
+				return nil, &Error{
+					full:  raw,
+					wrong: t,
+					desc:  "missing 'ON' after 'JOIN'",
+				}
+			}
+			subt := tokens[idx]
+			if _on.eq(subt) {
+				break
+			}
+			joinTokens = append(joinTokens, subt)
+		}
+		join := strings.Join(joinTokens, " ")
```

Comment:
The `WHERE` / `LIMIT` / `GROUP BY` statements will be stored in the last join, is it necessary to separate them?